File: app/src/main/java/com/bakkenbaeck/sol/widget/DashboardWidget.java
package com.bakkenbaeck.sol.widget;

import android.Manifest;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.widget.LinearLayout;
import android.widget.RemoteViews;

import com.bakkenbaeck.sol.R;
import com.bakkenbaeck.sol.service.SunsetService;
import com.bakkenbaeck.sol.util.UserVisibleData;
import com.bakkenbaeck.sunviewlib.SunView;

public class DashboardWidget extends AppWidgetProvider {

    static void updateAppWidget(final Context context,
                                final AppWidgetManager appWidgetManager,
                                final int appWidgetId) {
        registerForSunPhaseChanges(context);
        startSunsetService(context);
    }

    private static void startSunsetService(final Context context) {
        final Intent service = new Intent(context, SunsetService.class);
        service.putExtra(SunsetService.EXTRA_SHOW_NOTIFICATION, false);
        context.startService(service);
    }

    @Override
    public void onUpdate(final Context context,
                         final AppWidgetManager appWidgetManager,
                         final int[] appWidgetIds) {
        for (int appWidgetId : appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId);
        }
    }

    private static void registerForSunPhaseChanges(final Context context) {
        final SunsetBroadcastReceiver receiver = new SunsetBroadcastReceiver();
        final IntentFilter intentFilter = new IntentFilter(SunsetService.ACTION_UPDATE);
        intentFilter.addCategory(Intent.CATEGORY_DEFAULT);

        context.getApplicationContext().registerReceiver(receiver, intentFilter);
    }

    private static class SunsetBroadcastReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(final Context context, final Intent intent) {
            final UserVisibleData uvd = new UserVisibleData(context, intent);
            final RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.dashboard_widget);
            final AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
            final ComponentName componentName = new ComponentName(context, DashboardWidget.class);
            setTodaysMessage(context, views, uvd);
            setSunView(context, views, uvd);
            setLocationPermissionMessage(context, views, uvd);
            appWidgetManager.updateAppWidget(appWidgetManager.getAppWidgetIds(componentName), views);
        }

        private void setLocationPermissionMessage(
                final Context context,
                final RemoteViews views,
                final UserVisibleData uvd) {
            final boolean locationPermissionGranted = ActivityCompat.checkSelfPermission(
                    context,
                    Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED;
            final int visibility = locationPermissionGranted ? View.GONE : View.VISIBLE;
            final int secColor = uvd.getCurrentPhase().getSecondaryColor();
            views.setViewVisibility(R.id.location_message, visibility);
            views.setTextColor(R.id.location_message, ContextCompat.getColor(context, secColor));
        }

        private void setSunView(
                final Context context,
                final RemoteViews views,
                final UserVisibleData uvd) {
            final int viewWidth = context.getResources().getDimensionPixelSize(R.dimen.widget_width);
            final int viewHeight = context.getResources().getDimensionPixelSize(R.dimen.widget_height);

            final SunView sunView = createSunView(context, uvd, viewWidth, viewHeight);
            renderSunView(views, viewWidth, viewHeight, sunView);
        }

        private void setTodaysMessage(
                final Context context,
                final RemoteViews views,
                final UserVisibleData uvd) {
            final int secColor = uvd.getCurrentPhase().getSecondaryColor();
            views.setTextViewText(R.id.todays_message, uvd.getTodaysMessage());
            views.setTextColor(R.id.todays_message, ContextCompat.getColor(context, secColor));
        }

        private SunView createSunView(final Context context, final UserVisibleData uvd, final int viewWidth, final int viewHeight) {
            final int priColor = uvd.getCurrentPhase().getPrimaryColor();
            final SunView sunView = new SunView(context);
            sunView.setLayoutParams(new LinearLayout.LayoutParams(viewWidth, viewHeight));
            sunView.setColor(ContextCompat.getColor(context, priColor));
            sunView.setStartLabel(uvd.getSunriseText());
            sunView.setEndLabel(uvd.getSunsetText());
            sunView.setColor(ContextCompat.getColor(context, priColor))
                    .setStartLabel(uvd.getSunriseText())
                    .setEndLabel(uvd.getSunsetText())
                    .setPercentProgress(uvd.getProgress());
            return sunView;
        }

        private void renderSunView(final RemoteViews views, final int viewWidth, final int viewHeight, final SunView sunView) {
            sunView.measure(viewWidth, viewHeight);
            sunView.layout(0,0,viewWidth,viewHeight);
            sunView.setDrawingCacheEnabled(true);
            final Bitmap bitmap = sunView.getDrawingCache();
            views.setImageViewBitmap(R.id.sun_view, bitmap);
        }
    }
}

File: app/src/main/java/com/bakkenbaeck/sol/widget/DashboardWidget.kt
package com.bakkenbaeck.sol.widget

import android.Manifest
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.view.View
import android.widget.LinearLayout
import android.widget.RemoteViews
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.bakkenbaeck.sol.BaseApplication

import com.bakkenbaeck.sol.R
import com.bakkenbaeck.sol.util.UserVisibleData
import com.bakkenbaeck.sol.view.MainActivity
import com.bakkenbaeck.sol.view.custom.SunView

class DashboardWidget : AppWidgetProvider() {

    companion object {
        internal fun updateAppWidget(context: Context) {
            registerForSunPhaseChanges(context)
            BaseApplication.instance.refreshLocation(false)
        }

        private fun registerForSunPhaseChanges(context: Context) {
            val receiver = SunsetBroadcastReceiver()
            val intentFilter = IntentFilter(BaseApplication.ACTION_UPDATE)
            intentFilter.addCategory(Intent.CATEGORY_DEFAULT)

            context.applicationContext.registerReceiver(receiver, intentFilter)
        }
    }

    override fun onUpdate(context: Context,
                          appWidgetManager: AppWidgetManager,
                          appWidgetIds: IntArray) {
        for (appWidgetId in appWidgetIds) {
            updateAppWidget(context)
        }
    }

    private class SunsetBroadcastReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val uvd = UserVisibleData(context, intent)
            val views = RemoteViews(context.packageName, R.layout.dashboard_widget)
            val appWidgetManager = AppWidgetManager.getInstance(context)
            val componentName = ComponentName(context, DashboardWidget::class.java)
            setBackground(views, uvd)
            setTodaysMessage(context, views, uvd)
            setSunView(context, views, uvd)
            setLocationPermissionMessage(context, views, uvd)
            makeWidgetClickable(context, views)
            appWidgetManager.updateAppWidget(appWidgetManager.getAppWidgetIds(componentName), views)
        }

        private fun setBackground(views: RemoteViews,
                                  uvd: UserVisibleData) {
            val color = uvd.currentPhase.backgroundColor
            views.setInt(R.id.widget_container, "setBackgroundResource", color)
        }

        private fun setLocationPermissionMessage(
                context: Context,
                views: RemoteViews,
                uvd: UserVisibleData) {
            val locationPermissionGranted = ActivityCompat.checkSelfPermission(
                    context,
                    Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED
            val visibility = if (locationPermissionGranted) View.GONE else View.VISIBLE
            val secColor = uvd.currentPhase.secondaryColor
            views.setViewVisibility(R.id.location_message, visibility)
            views.setTextColor(R.id.location_message, ContextCompat.getColor(context, secColor))
        }

        private fun setSunView(
                context: Context,
                views: RemoteViews,
                uvd: UserVisibleData) {
            val viewWidth = context.resources.getDimensionPixelSize(R.dimen.widget_width)
            val viewHeight = context.resources.getDimensionPixelSize(R.dimen.widget_height)

            val sunView = createSunView(context, uvd, viewWidth, viewHeight)
            renderSunView(views, viewWidth, viewHeight, sunView)
        }

        private fun setTodaysMessage(
                context: Context,
                views: RemoteViews,
                uvd: UserVisibleData) {
            val secColor = uvd.currentPhase.secondaryColor
            views.setTextViewText(R.id.todays_message, uvd.todaysMessage)
            views.setTextColor(R.id.todays_message, ContextCompat.getColor(context, secColor))
        }

        private fun createSunView(context: Context,
                                  uvd: UserVisibleData,
                                  viewWidth: Int,
                                  viewHeight: Int): SunView {
            val primaryColor = ContextCompat.getColor(context, uvd.currentPhase.primaryColor)
            return SunView(context).apply {
                layoutParams = LinearLayout.LayoutParams(viewWidth, viewHeight)
                setColor(primaryColor)
                setStartLabel(uvd.sunriseText)
                setEndLabel(uvd.sunsetText)
                setColor(primaryColor)
                setPercentProgress(uvd.progress)
            }
        }

        private fun renderSunView(views: RemoteViews,
                                  viewWidth: Int,
                                  viewHeight: Int,
                                  sunView: SunView) {
            sunView.measure(viewWidth, viewHeight)
            sunView.layout(0, 0, viewWidth, viewHeight)
            sunView.isDrawingCacheEnabled = true
            val bitmap = sunView.drawingCache
            views.setImageViewBitmap(R.id.sun_view, bitmap)
        }

        private fun makeWidgetClickable(context: Context, views: RemoteViews) {
            val intent = Intent(context, MainActivity::class.java)
            val pendingIntent = PendingIntent.getActivity(context, 0, intent, 0)
            views.setOnClickPendingIntent(R.id.widget_container, pendingIntent)
        }
    }
}

